File: code/005/python/main.py
"""From Python to Go: Python: 005 - Code Flow Control: Conditionals and Loops"""

# Import

# Variables
# Your initial inventory
inventory = [
    ["leaf-01", "cisco-nxos", "192.168.1.1"],
    ["leaf-02", "arista-eos", "192.168.1.2"],
    ["spine-01", "cisco-nxos", ""],
    ["spine-02", "arosta-eos", "192.168.1.12"],
]


# Functions
def get_data(ip_address: str) -> list:
    """Function that pretends to connect to network device and collect some output"""
    # Ensure there is IP address to connect
    if not ip_address:
        return (False, "There is no IP address provided")

    # Return some mock data
    return (True, "Some raw data")


def parse_data(parser: str, data: str) -> list:
    """Function that pretends to parse the output from network device depeinding on its operating system type"""
    match parser:
        case "cisco-nxos":
            return (True, f"parsed: {data}")

        case "arista-eos":
            return (True, f"parsed: {data}")

        case _:
            return (False, "There is no parser available")


# Execution
if __name__ == "__main__":
    # Loop through all network devices
    for device in inventory:
        # Collect data for each network device
        collected_data = get_data(device[2])

        parsed_data = []
        # Do parsing if data is collected
        if collected_data[0]:
            parsed_data = parse_data(device[1], collected_data[1])
        else:
            print(f"Collecting data from {device[0]} is not successful")
            continue

        # Print results
        if parsed_data[0]:
            print(f"Successfully collected and parsed data for {device[0]}")
        else:
            print(f"Successfully collected but NOT parsed data for {device[0]}")

File: code/005/go/main.go
/* From Python to Go: Python: 005 - Code Flow Control */

package main

import "fmt"

// Aux functions
func getData(ipAddress string) []string {
	/* Function that pretends to connect to network device and collect some output */

	// Ensure there is IP address to connect
	if ipAddress == "" {
		return []string{"NOT OK", "There is no IP address provided"}
	}

	// Return some mock data
	return []string{"OK", "Some raw data"}
}

func parseData(parser, data string) []string {
	/* Function that pretends to parse the output from network device depeinding on its operating system type */

	switch parser {
	case "cisco-nxos":
		return []string{"OK", fmt.Sprintf("parsed: %v\n", data)}
	case "arista-eos":
		return []string{"OK", fmt.Sprintf("parsed: %v\n", data)}
	default:
		return []string{"NOT-OK", "There is no parser available"}
	}
}

// Main
func main() {
	// Define inventory
	inventory := [][]string{
		{"leaf-01", "cisco-nxos", "192.168.1.1"},
		{"leaf-02", "arista-eos", "192.168.1.2"},
		{"spine-01", "cisco-nxos", ""},
		{"spine-02", "arosta-eos", "192.168.1.12"},
	}

	// Loop through all network devices
	for _, device := range inventory {
		// Collect data for each network device
		collectedData := getData(device[2])

		var parsedData []string

		// Do parsing if data is collected
		if collectedData[0] == "OK" {
			parsedData = parseData(device[1], collectedData[1])
		} else {
			fmt.Printf("Collecting data from %v is not successful\n", device[0])
			continue
		}

		// Print results
		if parsedData[0] == "OK" {
			fmt.Printf("Successfully collected and parsed data for %v\n", device[0])
		} else {
			fmt.Printf("Successfully collected but NOT parsed data for %v\n", device[0])
		}
	}
}
